tests/MigrationAPI.cpp

```diff
@@ -0,0 +1,153 @@
+    REQUIRE(con.connection.use_count() == 2);
+
+    // This checks whether another connection is opened - which shouldn't because all make_storages() use the DbConnection sqlite3*
+    sqlite3* pDb = nullptr;
+    oldStorage.on_open = [&pDb](sqlite3* p) {
+        pDb =
+            p;  // must never be called because on_open_internal is only called once in DbConnection's constructor (since every further call uses an open shared connection)
+    };
+    oldStorage.drop_table(oldStorage.tablename<OldUser>());
+    oldStorage.sync_schema();
+    oldStorage.pragma.user_version(1);
+    oldStorage.replace(OldUser{1, "juan dent"});
+    oldStorage.replace(OldUser{2, "klaus"});
+    oldStorage.replace(OldUser{3, "eugene"});
+    retain_count = oldStorage.get_connection_holder()->retain_count();
+    REQUIRE(retain_count == 1);
+    REQUIRE(con.connection.use_count() == 2);
+    REQUIRE(pDb == nullptr);
+
+    SECTION("register_migrations") {
+        oldStorage.register_migration(1, 2, [](const DbConnection& con) {
+            internal::connection_ref{*con.connection};
+            std::map<int, std::string> emails = {
+                {1, "fnc12@me.com"},
+                {2, "megasuperhero@gmail.com"},
+                {3, "thor@hotmail.com"},
+            };
+            auto oldStorage = version1::getStorage(con);
+            auto retain_count = oldStorage.get_connection_holder()->retain_count();
+            REQUIRE(retain_count == 1);
+            auto allOldUsers = oldStorage.get_all<version1::User>();
+            auto nextStorage = version2::getStorage(con);
+            retain_count = nextStorage.get_connection_holder()->retain_count();
+            nextStorage.sync_schema();  // old users are dropped here
+            nextStorage.pragma.user_version(2);
+            retain_count = nextStorage.get_connection_holder()->retain_count();
+            REQUIRE(retain_count == 1);
+            for(auto& oldUser: allOldUsers) {
+                version2::User newUser{oldUser.id, move(oldUser.name), emails[oldUser.id]};
+                nextStorage.insert(newUser);
+            }
+            auto count = con.connection.use_count();
+        });
+        oldStorage.register_migration(2, 3, [](const DbConnection& con) {
+            internal::connection_ref{*con.connection};
+            auto oldStorage = version2::getStorage(con);
+            auto allOldUsers = oldStorage.get_all<version2::User>();
+            auto nextStorage = version3::getStorage(con);
+
+            nextStorage.sync_schema();  // old users are dropped here
+            nextStorage.pragma.user_version(3);
+
+            for(auto& oldUser: allOldUsers) {
+                auto& oldUserName = oldUser.name;
+                auto spaceIndex = oldUserName.find(' ');
+                version3::User newUser{oldUser.id};
+                if(spaceIndex != oldUserName.npos) {  // space is found
+                    auto firstName = oldUserName.substr(0, spaceIndex++);
```

Comment:
I bet `spaceIndex++` must be replaced with `spaceIndex++ - 1`

Reply:
no not really